src/bom/chainsop.rs

```diff
@@ -0,0 +1,872 @@
+    /// executing a command in a sub-process.  This is useful for interleaving
+    /// local processing into the command chain where that local processing is
+    /// executed in proper sequence with the other commands.  The local function
+    /// is provided with the "argument list" that would have been passed on the
+    /// command-line; this argument list will contain any input or output
+    /// filenames that should be used by the function.
+    ///
+    /// A local function execution in the chain can only pass an output file to
+    /// the subsequent operation in the chain; more complex data exchange would
+    /// need to be serialized into that output file and appropriately consumed by
+    /// the next stage. This might initially seem awkward, but makes sense when
+    /// you consider that most operations are executions in subprocesses that are
+    /// in a separate address space already.
+    pub fn calling<T>(f: T) -> SubProcOperation
+    where T: Fn(&Path, Vec<OsString>) -> anyhow::Result<()> + 'static
+    {
+        SubProcOperation {
+            cmd : Operation::Call(Box::new(f)),
+            args : Vec::new(),
+            inp_file : Default::default(),
+            out_file : Default::default(),
+            in_dir : None,
+        }
+    }
+
+    /// Adds a command-line argument to use when executing the command.
+    #[inline]
+    pub fn push_arg<T>(&mut self, arg: T) -> &SubProcOperation
+    where OsString: From<T>
+    {
+        self.args.push(OsString::from(arg));
+        self
+    }
+
+    /// Sets the input file for the command, overriding any previous input file
+    /// specification.
+    #[inline]
+    pub fn set_input_file(&mut self, inp_file: &FileSpec) -> &SubProcOperation
+    {
+        self.inp_file = inp_file.clone();
+        self
+    }
+
+    /// Sets the output file for the command, overriding any previous output file
+    /// specification.
+    #[inline]
+    pub fn set_output_file(&mut self, out_file: &FileSpec) -> &SubProcOperation
+    {
+        self.out_file = out_file.clone();
+        self
+    }
+
+    /// Sets the directory from which the operation will be executed.  The caller
+    /// is responsible for ensuring any Actual FileSpec paths are valid when
+    /// operating from that directory and any Temp FileSpec files created will
+    /// still be created in the normal temporary directory location.
+    #[inline]
+    pub fn set_dir<T>(&mut self, in_dir: T) -> &SubProcOperation
+    where T: AsRef<Path>
+    {
+        self.in_dir = Some(in_dir.as_ref().to_path_buf());
+        self
+    }
+
+    /// Executes this command in a subprocess in the specified directory.  The
+    /// input and output files will be determined and added to the command-line
+    /// as indicated by their FileSpec values.  The successful result specifies
+    /// the output file written (if any).
+    pub fn execute(&self, cwd: &Path) -> anyhow::Result<SubProcFile>
```

Comment:
How does this `cwd` parameter relate to `set_dir` above?

Reply:
Updated the documentation to describe this.